### DISCUSSION TITLE
Running metadata.reflect on my Aurora pgsql DB throws

### DISCUSSION THREAD

**Discussion by mat100payette** [sqlalchemy/sqlalchemy#11269] [Usage Questions] [answered]
Hello. Not sure yet if this is a bug so please enlighten me. When I run `metadata.reflect(bind=engine)` on my AWS Aurora PGSQL DB, I get the following cryptic error:

```
An error occurred: (aurora_data_api.exceptions.DatabaseError) An error occurred (BadRequestException) when calling the ExecuteStatement operation: ERROR: function generate_subscripts(smallint[], bigint) does not exist
  Hint: No function matches the given name and argument types. You might need to add explicit type casts.
  Position: 568; SQLState: 42883
[SQL: SELECT attr.conrelid, array_agg(CAST(attr.attname AS TEXT) ORDER BY attr.ord) AS cols, attr.conname, min(attr.description) AS description, NULL AS extra
FROM (SELECT con.conrelid AS conrelid, con.conname AS conname, con.conindid AS conindid, con.description AS description, con.ord AS ord, pg_catalog.pg_attribute.attname AS attname
FROM pg_catalog.pg_attribute JOIN (SELECT pg_catalog.pg_constraint.conrelid AS conrelid, pg_catalog.pg_constraint.conname AS conname, pg_catalog.pg_constraint.conindid AS conindid, unnest(pg_catalog.pg_constraint.conkey) AS attnum, generate_subscripts(pg_catalog.pg_constraint.conkey, :generate_subscripts_1) AS ord, pg_catalog.pg_description.description AS description
FROM pg_catalog.pg_constraint LEFT OUTER JOIN pg_catalog.pg_description ON pg_catalog.pg_description.objoid = pg_catalog.pg_constraint.oid      
WHERE pg_catalog.pg_constraint.contype = :contype AND pg_catalog.pg_constraint.conrelid IN (:oids_1, :oids_2, :oids_3, :oids_4, :oids_5, :oids_6, :oids_7, :oids_8, :oids_9, :oids_10, :oids_11, :oids_12, :oids_13, :oids_14, :oids_15, :oids_16, :oids_17)) AS con ON pg_catalog.pg_attribute.attnum = con.attnum AND pg_catalog.pg_attribute.attrelid = con.conrelid
WHERE con.conrelid IN (:oids_1, :oids_2, :oids_3, :oids_4, :oids_5, :oids_6, :oids_7, :oids_8, :oids_9, :oids_10, :oids_11, :oids_12, :oids_13, :oids_14, :oids_15, :oids_16, :oids_17)) AS attr GROUP BY attr.conrelid, attr.conname ORDER BY attr.conrelid, attr.conname]
[parameters: {'generate_subscripts_1': 1, 'contype': 'p', 'oids_1': 1687569, 'oids_2': 20638, 'oids_3': 1220625, 'oids_4': 98913, 'oids_5': 20490, 'oids_6': 20557, 'oids_7': 20542, 'oids_8': 20512, 'oids_9': 565265, 'oids_10': 1237069, 'oids_11': 647185, 'oids_12': 565288, 'oids_13': 98936, 'oids_14': 20572, 'oids_15': 20502, 'oids_16': 20620, 'oids_17': 20595}]
```

When I track it down, it seems to stem from the `get_multi_columns` function of `sqlalchemy\dialects\postgresql\base.py` where that query is built. I'm not sure what I'm doing that could cause this. The `sqlalchemy` version I'm using is `2.0.29` and my Aurora cluster uses `Postgres 13.12`.

**Comment by zzzeek** [comment]
what does your driver URL look like?

**Reply by CaselIT** [reply]
this may be a driver issue that uses int8 for a plain int passed as bound parameter.
PG in his infinite wisdom raises here
```py
SELECT generate_subscripts(ARRAY[1, 2, 4]::smallint[] , 1::int8)
```

We probably need to add an explicit type cast

**Reply by zzzeek** [reply]
yah this is why im asking about driver since they can just switch that for now

**Reply by CaselIT** [reply]
> We probably need to add an explicit type cast

or the driver in use needs to turn on casting, similar to what sqlalchemy does for asyncpg / psycopg3

**Reply by mat100payette** [reply]
Sorry for the late reply, here's my driver string:
```
"postgresql+auroradataapi://:@/my_db_name"
```

**Reply by CaselIT** [reply]
this looks like something that should be reported to the `auroradataapi` developers.
They should likely enable `render_bind_cast` similar to how psycopg does since it seems that that driver uses server side binding
https://github.com/sqlalchemy/sqlalchemy/blob/483edf651f6c3db6729f8d75265241b67ccd6b80/lib/sqlalchemy/dialects/postgresql/psycopg.py#L120-L121

not having them is simply asking for errors given how picky postgresql is with types

@zzzeek do you think adding a cast on the query is still something we should do?

**Reply by zzzeek** [reply]
i dont think so yet until we know more about this.   like if the user here wanted to try out psycopg or pg8000 with their aurora DB so we can get a better handle on what's happening.   is auroradataapi using a DBAPI as a dependency or is it itself directly a DBAPI?

**Reply by mat100payette** [reply]
> i dont think so yet until we know more about this. like if the user here wanted to try out psycopg or pg8000 with their aurora DB so we can get a better handle on what's happening. is auroradataapi using a DBAPI as a dependency or is it itself directly a DBAPI?

I'm willing to try different things if it means I can get away from `aurora-data-api` (which is paired with `sqlalchemy-aurora-data-api`). What are you suggesting I try? My main usage of those libs is to generate the schema of my Aurora db and save that in a file that I use in my own private libs.

**Reply by mat100payette** [reply]
After some looking around, I believe using those packages is probably unavoidable? Amazon RDS DataAPI doesn't seem to have any other sqlalchemy dialect than the one they provide unfortunately. What exactly did you mean with _"if the user here wanted to try out psycopg or pg8000 with their aurora DB"_?

**Reply by CaselIT** [reply]
I'm not 100% sure what `Amazon RDS DataAPI` is.
Is it just a managed postgresql? If so you should be able to use any postgresql dialect, like psycopg or pg8000.
If it's instead an amanon proprietary thing that's kind compatible with PG but not really PG, you are likely forced to use their stuff. And since you pay for it you should ask for guidance to the aws team that develops the aurora dbapi (assuming it's a community project)

**Reply by mat100payette** [reply]
That's the full dialect code for it:
[https://github.com/cloud-utils/sqlalchemy-aurora-data-api/blob/main/sqlalchemy_aurora_data_api/__init__.py](https://github.com/cloud-utils/sqlalchemy-aurora-data-api/blob/main/sqlalchemy_aurora_data_api/__init__.py)

I have a fork of it so I can make changes. Would the `render_bind_cast` suggestion you gave be applicable there?

**Reply by CaselIT** [reply]
It should help.

**Reply by mat100payette** [reply]
Unfortunately, adding these classes didn't work and I'm still getting the same exact error.

Here's what I added to the dialect code:

```py
class _ADA_STRING(sqltypes.String): 
    render_bind_cast = True

class _PGBigInteger(sqltypes.BigInteger):
    render_bind_cast = True

class _PGSmallInteger(sqltypes.SmallInteger):
    render_bind_cast = True

# ...

colspecs = util.update_copy(
    PGDialect.colspecs,
    {
        # ...
        sqltypes.String: _ADA_STRING,
        sqltypes.SmallInteger: _PGSmallInteger,
        sqltypes.BigInteger: _PGBigInteger,
    },
)
```

If you have any other ideas I can try please let me know; this is preventing me from generating the schema for my database.

**Reply by CaselIT** [reply]
well if you have a custom dialect the easiest whay would be to modify the query that's causing an issue.
second option could be to create that function, calling to the default `generate_subscripts` with a cast to int for the second arg

**Reply by mat100payette** [reply]
Ayaya, I wish I had any idea what your suggestion involves but I honestly don't know where to add/change stuff. I have very limited understanding of how the dialects and custom dbapi clients work. When you say "modify the query that's causing an issue", what exactly does that entail? When I follow the debugger to the query creation, I get to the `_columns_query` function of `sqlalchemy\dialects\postgresql\base.py` which isn't my code. Sorry for the lack of experience though.

**Reply by CaselIT** [reply]
oh I thought you were using a custom dialect that subclassed the default postgresql one.

**Reply by mat100payette** [reply]
I'll just drop the two files that matter in here, let me know if there's anything to be done there.
[sqlalchemy_dataapi.zip](https://github.com/sqlalchemy/sqlalchemy/files/15046219/sqlalchemy_dataapi.zip)

**Reply by mat100payette** [reply]
So I found that the `generate_subscripts` sql function is added in the `_constraint_query` function of `PGDialect` (which at runtime is my `AuroraPostgresDataAPIDialect` object/class). What's a proper way to patch that? Do I copy that `_constraint_query` function in my `AuroraPostgresDataAPIDialect` class and do the modifications there or is there a cleaner way to approach that?

**Reply by mat100payette** [reply, marked as answer]
Sorry for the multiple replies. I've got somewhere with this, enough to have the query run, but I'm not sure if what I've done is optimal or even correct. Here are the changes I applied:

1. Copied the following functions of `sqlalchemy.dialects.postgresql.base` into my own `AuroraPostgresDataAPIDialect` class: [`_constraint_query`](https://github.com/sqlalchemy/sqlalchemy/blob/7adc7404acc691698e30c362a8ec03af2bd426fd/lib/sqlalchemy/dialects/postgresql/base.py#L3973), [`_index_query`](https://github.com/sqlalchemy/sqlalchemy/blob/7adc7404acc691698e30c362a8ec03af2bd426fd/lib/sqlalchemy/dialects/postgresql/base.py#L4347) and [`get_multi_indexes`](https://github.com/sqlalchemy/sqlalchemy/blob/7adc7404acc691698e30c362a8ec03af2bd426fd/lib/sqlalchemy/dialects/postgresql/base.py#L4486).
2. In  `_constraint_query`, I changed the following:
    ```py
    sql.func.generate_subscripts(
        pg_catalog.pg_constraint.c.conkey, 1
    ).label("ord"),
    ```
    to
    ```py
    sql.func.generate_subscripts(
        pg_catalog.pg_constraint.c.conkey, sql.cast(1, INTEGER)
    ).label("ord"),
    ```
3. In `_index_query `, I applied the same `cast` in the `generate_subscripts` call, plus the following change:
    ```py
    pg_catalog.pg_get_indexdef(
        idx_sq.c.indexrelid, idx_sq.c.ord + 1, True
    ),
    ```
    to
    ```py
    pg_catalog.pg_get_indexdef(
        idx_sq.c.indexrelid, sql.cast(idx_sq.c.ord + 1, INTEGER), True
    ),
    ```
4. Finally in `get_multi_indexes`, I added this check:
    ```py
    for col_index, col_flags in enumerate(row["indoption"]):
        col_sorting = ()
        # try to set flags only if they differ from PG
        # defaults...
        try:
            col_flags = int(col_flags)
        except ValueError:
            continue
    
        if col_flags & 0x01:
            ...
    ```
    to
    ```py
    for col_index, col_flags in enumerate(row["indoption"]):
        col_sorting = ()
        # try to set flags only if they differ from PG
        # defaults...
    
        if col_flags & 0x01:
            ...
    ```

The casts to `INTEGER` were done because the `1` was showing up as `bigint` in the final query but postgresql doesn't seem to have a signature for `generate_subscripts` and `pg_get_indexdef` that take in `bigint`, just `int`.

The added check was done because `col_flags` in the debugger was a string (like `'0'`) and sometimes even just a space `' '`. I didn't check more in details why and/or if that is normal, but I hope that this can help you guys determine if I should be good with this or if there's something more needed on my and/or your end. 🙏

**Reply by CaselIT** [reply]
I still think the issue here is in your driver. It's not correct to threat 1 as bigint, considering how sensitive is postgresql.

The change you have done seem fine, feel free to propose ther to the AuroraPostgresDataAPIDialect dialect

**Reply by mat100payette** [reply]
The only thing I can see would maybe be this in the driver?

```py
_data_api_type_map = {
      bytes: "blobValue",
      bool: "booleanValue",
      float: "doubleValue",
      int: "longValue", # <-----
      str: "stringValue",
      Decimal: "stringValue"
      # list: "arrayValue"
  }
```
But I don't see anything related to "bigint"

**Reply by CaselIT** [reply]
you should likely ask for guidance to the `AuroraPostgresDataAPIDialect ` developers

**Reply by mat100payette** [reply]
That is unfortunately not possible in my case. But if that's not something you guys can help with that's fine, I'll keep chipping at it ^^ Thanks

**Reply by CaselIT** [reply]
> That is unfortunately not possible in my case.

How come?

Feel free to keep asking here but I've never used that dialect, so it's likely that the support we can provide is limited.

**Reply by mat100payette** [reply]
In short, a long time ago I tried to help fixing some bugs for them and opened an issue and PRs. He asked me to make tests for it, but when I replied twice that I didn't know how and that there were no similar tests done in the repo for me to look at. The second time was a bit blunt, and he simply blocked me from his whole stuff. Such is life. He also doesn't really maintain it much according to the commits anyways.

So I'm maintaining my own private version of it. In any case, sorry for the hassle as most of this is new to me and dialects/drivers aren't really trivial haha. That being said, I'm curious to know why/if you think the fixes I made should be in my dialect and not fixes in the base dialect?

**Reply by CaselIT** [reply]
> In short,...

That's unfortunate.

>  That being said, I'm curious to know why/if you think the fixes I made should be in my dialect and not fixes in the base dialect?

I'm leaning towards a no. The driver or the aws service you are using seems to take very strange decisions regarding how to communicate to PostgreSQL, and the best way of accommodating for it is doing these changes in a 3rd party dialect

**Reply by mat100payette** [reply]
Okok! Well Thanks a lot for the quick replies, that was very much appreciated. I will mark as answered on the specific solution I found just for future reference but your replies helped! Have a nice day.
